### PR TITLE
Updated chart to be a stateful set.

### PR DESCRIPTION
**Pull request by gaigepr**
**What this PR does / why we need it**:

This PR updates the postgresql helm chart to use a stateful set instead of a deployment.  

**Which issue this PR fixes** : references  #1863 

**Special notes for your reviewer**: 

None.

### CODE REVIEW

**Comment by gaigepr** [comment]
@databus23

**Comment by gaigepr** [comment]
signed

File: stable/postgresql/Chart.yaml [hunk 1]
```diff
@@ -14,4 +14,5 @@ sources:
 - https://github.com/docker-library/postgres
 maintainers:
 - name: databus23
+- name: gaigepr
```
Comment (hiddeco) [inline stable/postgresql/Chart.yaml]: Are you sure you're a maintainer and not just rolling out a patch?

**Review by hiddeco — COMMENTED** [review]
I think it would be wise to also refactor the persistent volume to a `volumeClaimTemplate` (https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#components). This way the `templates/pvc.yaml` should become obsolete.

**Comment by gaigepr** [comment]
@hiddeco 
I have refactored the `statefulset` to use a `volumeClaimTemplate` for postgresql's data. I have maintained the `pvc` because it seems to be required for the metrics container that can optionally run with postgresql. Is this acceptable?

File: stable/postgresql/templates/statefulset.yaml [hunk 2]
```diff
@@ -8,10 +8,29 @@ metadata:
...
 spec:
+  serviceName: {{ template "postgresql.fullname" . }}
+  selector:
+    matchLabels:
+      app: {{ template "postgresql.fullname" . }}
   template:
     metadata:
       labels:
         app: {{ template "postgresql.fullname" . }}
+    volumeClaimTemplates:
```
Comment (gaigepr) [inline stable/postgresql/templates/statefulset.yaml:19]: Maintains use of the `values.yaml` file for configuring aspects of the `volumeClaimTemplate`

**Comment by gaigepr** [comment]
@hiddeco 
 Roadblocks:
- [ ] removing the `pvc` and the `volumes` sections of the `statefulset.yaml` file results in a `helm install .` to never create a pod. 
- [ ] Should the optional metrics container be a part of a vanilla postgresql helm chart? Should it have its own directory called postgres-prometheus?

File: stable/postgresql/templates/statefulset.yaml [hunk 3]
```diff
@@ -104,14 +123,6 @@ spec:
 {{ toYaml .Values.metrics.resources | indent 10 }}
 {{- end }}
       volumes:
```
Comment (gaigepr) [inline stable/postgresql/templates/statefulset.yaml:125]: removing this results in the following message from `kubectl get events` even thought the `volumeClaimTemplate` uses the name `data`.

```
Warning FailedCreate  statefulset-controller  
create Pod p1-postgresql-0 in StatefulSet p1-postgresql failed 
error: Pod "p1-postgresql-0" is invalid: spec.containers[0].volumeMounts[0].name: Not found: "data"
```

**Comment by unguiculus** [comment]
Why would it make sense to convert this to a statefulset?

**Comment by lucasyvas** [comment]
Maybe not directly related, but is this in preparation for clustering/replicas?

**Comment by gaigepr** [comment]
@unguiculus It is my understanding that databases in general should be statefulsets unless the application does not need durability guarantees. If you think this is not worthy of modification to the existing chart, would it instead be worth creating a new chart?

**Comment by gaigepr** [comment]
@lucasyvas I am not sure what you are talking about. I desire to have a statefulset chart for postgresql.

**Comment by unguiculus** [comment]
@gaigepr @lucasyvas It depends on the database whether or not a statefulset makes sense. This PR would create a PVC for each replica if you scale up. This is not how Postgres works. Statefulsets make sense for databases that can scale horizontally, such as Cassandra, CockroachDB, MongoDB, etc.

**Pull request closed without merging**